### DISCUSSION TITLE
Correct way to assert.Equal structs that contain functions pointers

### DISCUSSION THREAD

**Discussion by bLuka** [stretchr/testify#1686] [Q&A] [answered]
Hello,

I'm having troubles comparing objects indirectly referencing function pointers.

For the reference, here's a minimal example:

```go
package main

import (
        "testing"

        assert "github.com/stretchr/testify/assert"
)

type s struct {
        f func()
}

func TestAssert(t *testing.T) {
        var f = func() {}
        var obj = s{f}

        assert.Equal(t, obj, obj)
}
```

Testing this code leads to this error:
```
--- FAIL: TestAssert (0.00s)
    main_test.go:17: 
                Error Trace:    main_test.go:17
                Error:          Not equal: 
                                expected: main.s{f:(func())(0x571380)}
                                actual  : main.s{f:(func())(0x571380)}
                            
                                Diff:
                Test:           TestAssert
```

In my specific case, this function pointer field comes from a private nested structure I don't have the ability to manipulate.

**Is there a way to compare structures indirectly embedding a function pointer?**

**Comment by brackendawson** [comment]
As is mentioned in equal's [docs](https://pkg.go.dev/github.com/stretchr/testify/assert#Equal) functions are only equal if they are both nil. Looking in the source of the reflect package, any further comparison is [apparently](https://pkg.go.dev/reflect#DeepEqual) not possible in Go.

Unfortunately it looks like you'll need to compare each non-function field individually. If you need to check the function fields then you'll probably need to test code which calls them for state.

**Comment by bLuka** [comment]
> As is mentioned in equal's [docs](https://pkg.go.dev/github.com/stretchr/testify/assert#Equal) functions are only equal if they are both nil. Looking in the source of the reflect package, any further comparison is [apparently](https://pkg.go.dev/reflect#DeepEqual) not possible in Go.

From what I got, it is not possible to just ignore function pointers, then? I saw `github.com/google/go-cmp/cmp` allows this by passing options. Is there an equivalent using only `testify`?

By the way, #535 mentions switching from `reflect` to `go-cmp`. It could also solve the issue.

> Unfortunately it looks like you'll need to compare each non-function field individually.

It's quite complicated in this very specific case, where the function pointer appears in a 3-levels nested private struct.

**Reply by dolmen** [reply]
Switching to `go-cmp` is a no-go: we don't want to add a dependency.

**Comment by brackendawson** [comment]
It will definitely be worth considering this case when switching to go-cmp, so thanks for linking the issues, hopefully that's picked up. It would almost certainly be something for a testify v2 release, which has been proposed but I'm not sure it's coming any time soon as the maintainers of this project seem quite stretched.

I'm fairly certain there isn't a way to do this using only testify. 😔

**Comment by OscarVanL** [comment]
I've been having the same problem.

**Comment by mauriciosl** [comment]
bump
In my case the test spec has an Interface, so I can't manually compare the fields.

**Comment by mauriciosl** [comment]
> bump In my case the test spec has an Interface, so I can't manually compare the fields.

In case anyone has the same problem, I ended up testing the type and then comparing the fields manually

**Comment by darkprof83** [comment]
> Is there a way to compare structures indirectly embedding a function pointer?

```go
require.Equal(fmt.Sprint(tt.wantFunc), fmt.Sprint(actFunc))
```
solved my problem
```go
//go:build testmode
package mypackagetype 

// need flag
// -tags testmode

type stringFunc struct {
  i int
  f string
}

func NewStringFunc(orig origFunc) stringFunc {
  return stringFunc{i: orig.i, f: fmt.Sprint(orig.f)}
}
```

```go
require.Equal(NewStringFunc(tt.want), NewStringFunc(act))
```

**Comment by oneumyvakin** [comment, marked as answer]
Format to string works for me:
```
wantStr := fmt.Sprintf("%#v", tt.want)
gotStr := fmt.Sprintf("%#v", got)
assert.Equal(t, wantStr, gotStr)
```
